File: src/wings/routing/manager.py
"""API pool manager — global pool + per-task-type score masks.

Each agent, skill, etc. manages a score mask over the global pool.
An API can be disabled (-inf), promoted (+delta), or demoted (-delta).
Selection uses softmax(effective_scores).
"""

import threading
from copy import deepcopy

from wings.routing.types import PoolConfig, PoolEntry, ScoreMask
from wings.routing.selector import NEG_INF, NoAPIAvailable, softmax_select


class APIPoolManager:
    """Manages a global pool of APIs with per-task-type score masks.

    Implements ModelSelector Protocol.
    Thread-safe: all writes hold an RLock.
    """

    def __init__(self, config: PoolConfig | None = None):
        self._lock = threading.RLock()
        self._entries: dict[str, PoolEntry] = {}  # api_id → PoolEntry (global)
        self._masks: dict[str, dict[str, float]] = {}  # task_type → {api_id → delta}
        if config is not None:
            self.replace_config(config)

    # -- Selection (implements ModelSelector) --

    def select(self, task_type: str, override: str | None = None) -> str:
        """Select an API via softmax over effective scores. override bypasses."""
        if override:
            return override
        with self._lock:
            mask = self._resolve_mask(task_type)
            entries = list(self._entries.values())
            if not entries:
                raise NoAPIAvailable("no APIs in global pool")
            return softmax_select(entries, mask)

    # -- Registration --

    def register_api(
        self,
        api_id: str,
        *,
        score: float = 0.0,
    ) -> None:
        """Add an API to the global pool. Does NOT touch masks (effectively
        available to all task types at base score)."""
        with self._lock:
            self._entries[api_id] = PoolEntry(api_id=api_id, score=score)

    def unregister_api(self, api_id: str) -> None:
        """Remove an API from the global pool and all masks."""
        with self._lock:
            self._entries.pop(api_id, None)
            for mask in self._masks.values():
                mask.pop(api_id, None)

    # -- Score adjustments (per task type) --

    def _ensure_mask(self, task_type: str) -> dict[str, float]:
        if task_type not in self._masks:
            self._masks[task_type] = {}
        return self._masks[task_type]

    def adjust_score(self, task_type: str, api_id: str, score: float) -> None:
        """Set the score adjustment for an API in a task type's mask."""
        with self._lock:
            self._ensure_mask(task_type)[api_id] = score

    def adjust_base_score(self, api_id: str, score: float) -> None:
        """Set the base score of an API (affects all task types)."""
        with self._lock:
            if api_id not in self._entries:
                raise KeyError(api_id)
            self._entries[api_id].score = score

    def upvote(self, task_type: str, api_id: str, delta: float = 0.5) -> None:
        """Increase an API's score adjustment for a task type."""
        with self._lock:
            mask = self._ensure_mask(task_type)
            current = mask.get(api_id, 0.0)
            mask[api_id] = current + delta

    def downvote(self, task_type: str, api_id: str, delta: float = 0.5) -> None:
        """Decrease an API's score adjustment for a task type."""
        with self._lock:
            mask = self._ensure_mask(task_type)
            current = mask.get(api_id, 0.0)
            mask[api_id] = current - delta

    def disable(self, task_type: str, api_id: str) -> None:
        """Disable an API for a task type (mask delta = -inf)."""
        with self._lock:
            self._ensure_mask(task_type)[api_id] = NEG_INF

    def enable(self, task_type: str, api_id: str) -> None:
        """Re-enable a previously disabled API (remove mask entry)."""
        with self._lock:
            if task_type in self._masks:
                self._masks[task_type].pop(api_id, None)

    # -- Pool fork --

    def fork_mask(self, task_type: str, from_task: str) -> None:
        """Copy another task type's mask as a starting point."""
        with self._lock:
            source = self._masks.get(from_task, {})
            self._masks[task_type] = deepcopy(source)

    # -- Query --

    def list_apis(self) -> list[str]:
        """List all APIs in the global pool."""
        with self._lock:
            return sorted(self._entries.keys())

    def get_mask(self, task_type: str) -> dict[str, float]:
        """Get the effective mask for a task type (resolved via inheritance)."""
        with self._lock:
            return dict(self._resolve_mask(task_type))

    def list_task_types(self) -> list[str]:
        """List task types that have explicit masks."""
        with self._lock:
            return sorted(self._masks.keys())

    # -- Persistence --

    def to_config(self) -> PoolConfig:
        """Export global pool + masks to PoolConfig."""
        with self._lock:
            return PoolConfig(
                version=2,
                apis=[deepcopy(e) for e in self._entries.values()],
                masks={
                    tt: dict(m)
                    for tt, m in self._masks.items()
                    if m  # only non-empty masks
                },
            )

    def replace_config(self, config: PoolConfig) -> None:
        """Replace state from a loaded PoolConfig."""
        with self._lock:
            self._entries.clear()
            self._masks.clear()
            for entry in config.apis:
                self._entries[entry.api_id] = deepcopy(entry)
            for task_type, adjustments in config.masks.items():
                self._masks[task_type] = dict(adjustments)

    # -- Internal --

    def _resolve_mask(self, task_type: str) -> dict[str, float]:
        """Resolve mask via inheritance chain. Falls back to parent masks."""
        from wings.routing.tasks import resolve_parent

        # Walk up the chain collecting non-empty masks
        current = task_type
        resolved: dict[str, float] = {}
        while current is not None:
            mask = self._masks.get(current, {})
            for api_id, delta in mask.items():
                if api_id not in resolved:
                    resolved[api_id] = delta
            current = resolve_parent(current)
        return resolved

File: src/routing/manager.ts
/** API pool manager — global pool + per-task-type score masks.
 *
 * Each agent, skill, etc. manages a score mask over the global pool.
 * An API can be disabled (-Infinity), promoted (+delta), or demoted (-delta).
 * Selection uses softmax(effective_scores).
 *
 * No locks needed: JS is single-threaded.
 */

import type { ModelSelector } from "./protocol.ts";
import type { PoolConfig, PoolEntry } from "./types.ts";
import { NEG_INF, NoAPIAvailable, softmaxSelect } from "./selector.ts";
import { resolveParent } from "./tasks.ts";

export class APIPoolManager implements ModelSelector {
  /** api_id → PoolEntry (global). */
  private _entries: Map<string, PoolEntry> = new Map();
  /** task_type → {api_id → delta}. */
  private _masks: Map<string, Map<string, number>> = new Map();

  constructor(config?: PoolConfig) {
    if (config) this.replaceConfig(config);
  }

  // -- Selection (implements ModelSelector) --

  select(taskType: string, override?: string | null): string {
    if (override) return override;
    const mask = this._resolveMask(taskType);
    const entries = [...this._entries.values()];
    if (entries.length === 0) {
      throw new NoAPIAvailable("no APIs in global pool");
    }
    return softmaxSelect(entries, mask);
  }

  // -- Registration --

  /** Add an API to the global pool. Does NOT touch masks (effectively
   * available to all task types at base score). */
  registerApi(apiId: string, opts: { score?: number } = {}): void {
    this._entries.set(apiId, { api_id: apiId, score: opts.score ?? 0 });
  }

  /** Remove an API from the global pool and all masks. */
  unregisterApi(apiId: string): void {
    this._entries.delete(apiId);
    for (const mask of this._masks.values()) mask.delete(apiId);
  }

  // -- Score adjustments (per task type) --

  private _ensureMask(taskType: string): Map<string, number> {
    let m = this._masks.get(taskType);
    if (!m) {
      m = new Map();
      this._masks.set(taskType, m);
    }
    return m;
  }

  /** Set the score adjustment for an API in a task type's mask. */
  adjustScore(taskType: string, apiId: string, score: number): void {
    this._ensureMask(taskType).set(apiId, score);
  }

  /** Set the base score of an API (affects all task types). */
  adjustBaseScore(apiId: string, score: number): void {
    const entry = this._entries.get(apiId);
    if (!entry) throw new Error(`Unknown api_id: ${apiId}`);
    entry.score = score;
  }

  /** Increase an API's score adjustment for a task type. */
  upvote(taskType: string, apiId: string, delta: number = 0.5): void {
    const mask = this._ensureMask(taskType);
    mask.set(apiId, (mask.get(apiId) ?? 0) + delta);
  }

  /** Decrease an API's score adjustment for a task type. */
  downvote(taskType: string, apiId: string, delta: number = 0.5): void {
    const mask = this._ensureMask(taskType);
    mask.set(apiId, (mask.get(apiId) ?? 0) - delta);
  }

  /** Disable an API for a task type (mask delta = -Infinity). */
  disable(taskType: string, apiId: string): void {
    this._ensureMask(taskType).set(apiId, NEG_INF);
  }

  /** Re-enable a previously disabled API (remove mask entry). */
  enable(taskType: string, apiId: string): void {
    const mask = this._masks.get(taskType);
    if (mask) mask.delete(apiId);
  }

  // -- Pool fork --

  /** Copy another task type's mask as a starting point. */
  forkMask(taskType: string, fromTask: string): void {
    const source = this._masks.get(fromTask) ?? new Map();
    this._masks.set(
      taskType,
      new Map(source.entries()),
    );
  }

  // -- Query --

  /** Return effective scores for a task type: api_id → {base, delta, effective}. */
  getPoolInfo(taskType: string): Record<
    string,
    { base: number; delta: number; effective: number }
  > {
    const mask = this._resolveMask(taskType);
    const result: Record<string, { base: number; delta: number; effective: number }> = {};
    for (const [apiId, entry] of this._entries) {
      const delta = mask[apiId] ?? 0;
      result[apiId] = {
        base: entry.score,
        delta,
        effective: entry.score + delta,
      };
    }
    return result;
  }

  /** List all APIs in the global pool. */
  listApis(): string[] {
    return [...this._entries.keys()].sort();
  }

  /** Get the effective mask for a task type (resolved via inheritance). */
  getMask(taskType: string): Record<string, number> {
    return this._resolveMask(taskType);
  }

  /** List task types that have explicit masks. */
  listTaskTypes(): string[] {
    return [...this._masks.keys()].sort();
  }

  // -- Persistence --

  /** Export global pool + masks to PoolConfig. */
  toConfig(): PoolConfig {
    return {
      version: 2,
      apis: [...this._entries.values()].map((e) => ({ ...e })),
      masks: Object.fromEntries(
        [...this._masks.entries()]
          .filter(([, m]) => m.size > 0)
          .map(([tt, m]) => [tt, Object.fromEntries(m)] as const),
      ),
    };
  }

  /** Replace state from a loaded PoolConfig. */
  replaceConfig(config: PoolConfig): void {
    this._entries.clear();
    this._masks.clear();
    for (const entry of config.apis) {
      this._entries.set(entry.api_id, { ...entry });
    }
    for (const [taskType, adjustments] of Object.entries(config.masks)) {
      this._masks.set(taskType, new Map(Object.entries(adjustments)));
    }
  }

  // -- Internal --

  /** Resolve mask via inheritance chain. Falls back to parent masks. */
  private _resolveMask(taskType: string): Record<string, number> {
    let current: string | null = taskType;
    const resolved: Record<string, number> = {};
    while (current !== null) {
      const mask = this._masks.get(current);
      if (mask) {
        for (const [apiId, delta] of mask) {
          if (!(apiId in resolved)) resolved[apiId] = delta;
        }
      }
      current = resolveParent(current);
    }
    return resolved;
  }
}
